pkg/repository/maintenance.go

```diff
@@ -0,0 +1,233 @@
+/*
+Copyright the Velero contributors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package repository
+
+import (
+	"context"
+	"fmt"
+	"sort"
+	"time"
+
+	appsv1 "k8s.io/api/apps/v1"
+	batchv1 "k8s.io/api/batch/v1"
+	v1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"github.com/pkg/errors"
+
+	"github.com/vmware-tanzu/velero/pkg/repository/provider"
+	"github.com/vmware-tanzu/velero/pkg/util/kube"
+	"github.com/vmware-tanzu/velero/pkg/util/logging"
+	veleroutil "github.com/vmware-tanzu/velero/pkg/util/velero"
+)
+
+const RepositoryNameLabel = "velero.io/repo-name"
+const DefaultKeepLatestMaitenanceJobs = 3
+
+// MaintenanceConfig is the configuration for the repo maintenance job
+type MaintenanceConfig struct {
+	KeepLatestMaitenanceJobs int
+	CPURequest               string
+	MemRequest               string
+	CPULimit                 string
+	MemLimit                 string
+	LogLevelFlag             *logging.LevelFlag
+	FormatFlag               *logging.FormatFlag
+}
+
+func generateJobName(repo string) string {
+	millisecond := time.Now().UTC().UnixMilli() // millisecond
+
+	jobName := fmt.Sprintf("%s-maintain-job-%d", repo, millisecond)
+	if len(jobName) > 63 { // k8s job name length limit
+		jobName = fmt.Sprintf("repo-maintain-job-%d", millisecond)
+	}
+
+	return jobName
+}
+
+func buildMaintenanceJob(m MaintenanceConfig, param provider.RepoParam, cli client.Client, namespace string) (*batchv1.Job, error) {
```

Comment:
Can we copy the Velero pod directly and only remove or replace the specific needed fields (e.g. cmd)?
I raise this because the Velero pod may be configured in different ways to support kinds of scenarios, e.g. a label is added to it to support auth with Azure Workload Identity

Reply:
I missed the label settings for the job, I added them now.
As some properties are also used by other features later such as `data mover microservice`, so I want these properties to be a series of common functions.